File: js-template/leetcode/editor/cn/sliding-window-maximum.js
/*
 * @lc app=leetcode.cn id=239 lang=javascript
 * @lcpr version=30404
 *
 * [239] 滑动窗口最大值
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function(nums, k) {
    var s = [];

    var left = 0;
    var right = k - 1;

    var result = [];

    for (let i = 0; i < k; i++) {
        if (s.length === 0) {
            s.push(nums[i]);
        } else {
            while (s[s.length - 1] < nums[i]) {
                s.pop();
            }
            s.push(nums[i]);
        }
    }

    result.push(s[0]);

    for (let i = k; i < nums.length; i++) {
        const left = nums[i-k];
        
        if (left === s[0]) {
            s.shift();
        }

        while (s[s.length - 1] < nums[i]) {
            s.pop();
        }
        s.push(nums[i]);
        result.push(s[0]);
    }

    return result;
};
// @lc code=end

// your test code here


/*
// @lcpr case=start
// [1,3,-1,-3,5,3,6,7]\n3\n
// @lcpr case=end

// @lcpr case=start
// [1]\n1\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/sliding-window-maximum.cpp
/*
 * @lc app=leetcode.cn id=239 lang=cpp
 * @lcpr version=30404
 *
 * [239] 滑动窗口最大值
 */

#include <iostream>
#include <vector>
#include <string>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> cache;
        vector<int> result;
        int left = 0;

        for (int i = 0; i < nums.size(); ++i) {
            while (cache.size() > 0 && nums[cache.back()] < nums[i])
            {
                cache.pop_back();
            }
            cache.push_back(i);
            if (i - left + 1 >= k) {
                result.push_back(nums[cache.front()]);
                if (cache.front() == left)
                    cache.pop_front();
                ++left;
            }
        }
        return result;
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    vector<int> nums = {1,3,-1,-3,5,3,6,7};
    solution.maxSlidingWindow(nums, 3);
}



/*
// @lcpr case=start
// [1,3,-1,-3,5,3,6,7]\n3\n
// @lcpr case=end

// @lcpr case=start
// [1]\n1\n
// @lcpr case=end

 */

